keras_nlp/models/whisper/whisper_audio_feature_extractor.py

```diff
@@ -0,0 +1,222 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+
+NUM_MELS = 80
+
+
+class WhisperAudioFeatureExtractor(keras.layers.Layer):
+    """
+    Whisper audio feature extractor layer.
+
+    This layer takes in a batch of audio tensors, and computes the log-mel
+    spectrogram features for each audio tensor.
+
+    The input audio tensor can either be of shape `(length_of_audio,)` or
+    `(batch_size, length_of_audio)`. The output is a tensor of shape
+    `(batch_size, num_frames, NUM_MELS)`, where `num_frames` is
+    `length_of_audio / stride` and `NUM_MELS` is 80.
+
+    Args:
+        sample_rate: int, defaults to 16000. The sample rate of the audio.
+        num_fft_bins: int, defaults to 400. The size of the Fourier Transform in
+            STFT.
+        stride: int, defaults to 160. The distance between neighboring
+            sliding window frames while computing STFT.
+        chunk_length: int, defaults to 30. The length of each audio chunk in
+            seconds. The input audio tensor will be padded/trimmed to
+            `chunk_length*sample_rate`.
+    """
+
+    def __init__(
+        self,
+        sample_rate=16000,
+        num_fft_bins=400,
+        stride=160,
+        chunk_length=30,
+        **kwargs,
+    ):
+        # Check dtype and provide a default.
+        if "dtype" not in kwargs or kwargs["dtype"] is None:
+            kwargs["dtype"] = tf.float32
+        else:
+            dtype = tf.dtypes.as_dtype(kwargs["dtype"])
+            if not dtype.is_floating:
+                raise ValueError(
+                    f"dtype must be a floating type. Received: dtype={dtype}"
+                )
+
+        super().__init__(**kwargs)
+
+        self.sample_rate = sample_rate
+        self.num_fft_bins = num_fft_bins
+        self.stride = stride
+        self.chunk_length = chunk_length
+        self.n_samples = self.sample_rate * self.chunk_length
+
+        # After transposition, `self.mel_filters`'s shape is
+        # `(num_fft_bins // 2 + 1, NUM_MELS).`
+        self.mel_filters = tf.transpose(tf.constant(self._get_mel_filters()))
```

Comment:
somewhat awkward to add the transpose and constant here, why not just put these in the last two lines of the helper function?